storage/cache/subtree_cache.go

```diff
@@ -400,11 +405,23 @@ func PopulateMapSubtreeNodes(treeHasher merkle.TreeHasher) storage.PopulateSubtr
 // subtree Leaves map.
 //
 // This uses the CompactMerkleTree to repopulate internal nodes, and so will
-// handle imperfect (but left-hand dense) subtrees.
+// handle imperfect (but left-hand dense) subtrees. Note that we only rebuild internal
+// nodes when the subtree is fully populated. For an explanation of why see the comments
+// below for PrepareLogSubtreeWrite.
 func PopulateLogSubtreeNodes(treeHasher merkle.TreeHasher) storage.PopulateSubtreeFunc {
 	return func(st *storagepb.SubtreeProto) error {
-		st.InternalNodes = make(map[string][]byte)
 		cmt := merkle.NewCompactMerkleTree(treeHasher)
+		if st.Depth < 1 {
+			return fmt.Errorf("populate log subtree with invalid depth: %d", st.Depth)
+		}
+		fullyPopulatedLeafCount := 1 << uint(st.Depth)
+
+		// We're going to rebuild internal nodes if the subtree is fully populated so clear them now
+		if len(st.Leaves) == fullyPopulatedLeafCount {
+			st.InternalNodes = make(map[string][]byte)
+		}
+
```

Comment:
Maybe add a check that in the other case st.InternalNodes is not nil and contains a valid number of nodes given the number of leaves?